rcl/include/rcl/security.h

```diff
@@ -0,0 +1,104 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__SECURITY_H_
+#define RCL__SECURITY_H_
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <stdbool.h>
+
+#include "rcl/allocator.h"
+#include "rcl/types.h"
+#include "rcl/visibility_control.h"
+#include "rmw/security_options.h"
+
+#ifndef ROS_SECURITY_NODE_DIRECTORY_VAR_NAME
+  #define ROS_SECURITY_NODE_DIRECTORY_VAR_NAME "ROS_SECURITY_NODE_DIRECTORY"
+#endif
+
+#ifndef ROS_SECURITY_ROOT_DIRECTORY_VAR_NAME
+  #define ROS_SECURITY_ROOT_DIRECTORY_VAR_NAME "ROS_SECURITY_ROOT_DIRECTORY"
+#endif
+
+#ifndef ROS_SECURITY_LOOKUP_TYPE_VAR_NAME
+  #define ROS_SECURITY_LOOKUP_TYPE_VAR_NAME "ROS_SECURITY_LOOKUP_TYPE"
+#endif
+
+#ifndef ROS_SECURITY_STRATEGY_VAR_NAME
+#define ROS_SECURITY_STRATEGY_VAR_NAME "ROS_SECURITY_STRATEGY"
+#endif
+
+#ifndef ROS_SECURITY_ENABLE_VAR_NAME
+#define ROS_SECURITY_ENABLE_VAR_NAME "ROS_SECURITY_ENABLE"
+#endif
+
+/// Check if security has to be used, according to the environment.
+/**
+ * \param use_security[out] Must not be NULL.
+ * \returns RCL_RET_INVALID_ARGUMENT if an argument is not valid, or
+ * \returns RCL_RET_ERROR if an unexpected error happened, or
+ * \returns RCL_RET_OK.
+ */
+RCL_PUBLIC
+rcl_ret_t
+rcl_use_security(bool * use_security);
+
+/// Get security enforcement policy from the environment.
+/**
+ * \param policy[out] Must not be NULL.
+ * \returns RCL_RET_INVALID_ARGUMENT if an argument is not valid, or
+ * \returns RCL_RET_ERROR if an unexpected error happened, or
+ * \returns RCL_RET_OK.
+ */
+RCL_PUBLIC
+rcl_ret_t
+rcl_get_enforcement_policy(rmw_security_enforcement_policy_t * policy);
+
+/// Return the secure root given a name and namespace.
+/**
+ * The returned security directory is associated with the node or context depending on the
+ * rmw implementation.
+ *
+ * E.g. for a node/context named "c" in namespace "/a/b", the secure root path will
+ * be "a/b/c", where the delimiter "/" is native for target file system (e.g. "\\" for _WIN32).
+ * If no exact match is found for the node/context name, a best match would be used instead
+ * (by performing longest-prefix matching).
+ *
+ * However, this expansion can be overridden by setting the secure node directory environment
+ * variable (only in rmw implementations that associate a security directory with a node),
```

Comment:
I think that it also didn't make much sense for nodes.
I mean, you could pass a different `ROS_SECURITY_DIRECTORY` to each process, and that would work too (the files would need to be in `$ROS_SECURITY_DIRECTORY/node_namespace/node_name` instead of `ROS_SECURITY_NODE_DIRECTORY`).
If the intention was to specify the same security directory for many nodes in the same process, it wouldn't make much sense to do the same for context, as you can just use the same context name (node names are supposed to be unique).

In general, I try to avoid features that can cause confusion.
Using `ROS_SECURITY_NODE_DIRECTORY` when you have one `Participant` per `Context` would be confusing. A name for the environment variable like `ROS_SECURITY_CONTEXT_DIRECTORY` would make more sense, but I don't see the value of such an option.

Reply:
Yeah I agree, just makes me nervous to remove stuff I don't think makes sense, as I may not fully understand the problem the feature is trying to solve. Maybe @ruffsl or @mikaelarguedas will have some insight to share, here.

Reply:
Yeah I agree, just makes me nervous to remove stuff I don't think makes sense, as I may not fully understand the problem the feature is trying to solve. Maybe @ruffsl or @mikaelarguedas will have some insight to share, here.

Reply:
The main use I have for this is when I want to apply a single set of artifacts/policies to an entire subsystem.
e.g. "let's run this entire system with encryption and no access control and see how it performs"

Without this option we'd need to set a different environment with an appropriate env variable for each and every process that your launchfile (and included launchfiles) will start (that means a lot of launchfile editing)

Another example is to have an "admin" set of artifacts that can publish/subscribe to everything. You can point all your debugging tools to this single directory and not have to remap all your debug nodes / contexts to have the same name or prefix to fit the matching criteria.

---

So overall, even if it's not meant for use by default, I'm in favor of keeping such a feature

Reply:
> You can point all your debugging tools to this single directory and not have to remap all your debug nodes / contexts to have the same name or prefix to fit the matching criteria.

I think this use-case might be better solved by having the debugging tools setup to use a common context name that can get a set of predefined, wide-open profiles (this can be disabled for production, of course). @SidFaber has done a lot of thinking about that.

Reply:
> Without this option we'd need to set a different environment with an appropriate env variable for each and every process that your launchfile (and included launchfiles) will start (that means a lot of launchfile editing)

Maybe we could rename the variable to `ROS_SECURITY_DIRECTORY_OVERRIDE`, and avoid using `NODE` on it. We should keep backwards compatibility in rmw implementations that can actually set a different security directory for each point, and still accept the old name.

There's also another way to solve this. 99% of the processes, will have just one `Context` per process and use the default name on it. In that case, all launched processes will use the same security artifacts/policies. The only exception when you wouldn't use the default name, is when you need to create more than one (bridge between different domains, and other super stranges use cases). In those cases, you actually need to use a different context name, if not you won't be able to remap them separatly. For those exceptions, if you want to use the same security artifacts, you will have to remap both names againt to the "default context name".

Reply:
@kyrofa @mikaelarguedas Friendly ping.

Reply:
Sorry for the delay @ivanpauno, we've been chatting about exactly this in the WG. I promise we're not ignoring you!

Reply:
The general feeling is that this variable does indeed serve a purpose. I suggest keeping it, although I'm in favor of @ivanpauno's suggested rename.

Reply:
> you could pass a different ROS_SECURITY_DIRECTORY to each process

Having to specify an entirely different keystore could be tedious in the sense that one would have to take care that the two keystores shared the correct common certificate authorities and governances. I guess that isn't too difficult do one off, would currently consist of copying over the root level keystore files; though keeping the two keystores in sync would be fragile.

> I'm not sure if somebody was using it before

Here are some links to its purpose and origin:

* https://github.com/ros2/sros2/issues/69
* https://github.com/ros2/rcl/pull/300
* a82793ce3c70a9dc80fada177d01ac9181edfea8

> If the intention was to specify the same security directory for many nodes in the same process, it wouldn't make much sense to do the same for context, as you can just use the same context name
>
> ---
> We should keep backwards compatibility in rmw implementations that can actually set a different security directory for each point, and still accept the old name.

Hmm, I think I may have been under the impression that mapping multiple nodes per process to one context would remain optional, thus still warenting the use of `ROS_SECURITY_NODE_DIRECTORY`. However, reading further into this PR, it seems that the old style of mapping has been removed entirely (which I'm ok with). This may make it tricky to measure the minimal span of permissions required for each node that are never launched as their own process, as permissions brought in by other nodes could implicitly fill in the gaps of other node policies, akin to underspecified dependencies in package.xml manifest. But perhaps this just necessitates better measurement tooling of requested graph resources in SROS.

> There's also another way to solve this. 99% of the processes, will have just one Context per process and use the default name on it.

To be clear, while the norm may be for most processes to consist of only one context, different processes should not necessarily share a `default` context directory. Unless of course the user is unconcerned about access control, or the granularity of concern is that of a robot, and not a particular process in the robot. From a security perspective, our goal is to minimize the the attack surface or privilege of each process in the graph to what is only essential for it's application.

> For those exceptions, if you want to use the same security artifacts, you will have to remap both names againt to the "default context name".

Could you clarify by what you mean here? When you have a bridging scenario of one process, consisting of two separate contexts, say for bridging over two different DDS domains? If the two DDS domains where mutually exclusive, e.g. the ranges of domains expected to apply different permissions didn't overlap, then one could just use one context (one set of DDS security artifacts) as the DDS policy schema has it's primary pivot of allowed/denied permissions on the domain range.

https://github.com/ros2/sros2/blob/fc777a2b6957cc7481341bd9f7f49e7538035e9b/sros2/test/policies/permissions.xml#L10-L12

Going back to @mikaelarguedas https://github.com/ros2/rcl/pull/515#discussion_r372643420 , I think as long as users have a method for manually overriding what context directory gets loaded via the environment, for prototyping, deguing, and auditing, then that would serve the same purpose.

Reply:
> Having to specify an entirely different keystore could be tedious in the sense that one would have to take care that the two keystores shared the correct common certificate authorities and governances. I guess that isn't too difficult do one off, would currently consist of copying over the root level keystore files; though keeping the two keystores in sync would be fragile.

:+1: 

> However, reading further into this PR, it seems that the old style of mapping has been removed entirely (which I'm ok with).

Yes, the implementations will be refactored to map a `Participant` to a `Context`.

> Could you clarify by what you mean here? When you have a bridging scenario of one process, consisting of two separate contexts, say for bridging over two different DDS domains? If the two DDS domains where mutually exclusive, e.g. the ranges of domains expected to apply different permissions didn't overlap, then one could just use one context (one set of DDS security artifacts) as the DDS policy schema has it's primary pivot of allowed/denied permissions on the domain range.

Yes, I was refering to bridging two domains. In that case, you need two Contexts. as you need one participant for each domain. If they can share the security artifacts, both of them could use the same context name.

> Going back to @mikaelarguedas #515 (comment) , I think as long as users have a method for manually overriding what context directory gets loaded via the environment, for prototyping, deguing, and auditing, then that would serve the same purpose.

I will add `ROS_SECURITY_DIRECTORY_OVERRIDE`, which is a clearer name.

Reply:
Solved in https://github.com/ros2/rcl/pull/515/commits/d284a2e92ac07c3afc362b802d8b2851e302fbec.